// @sfast/gql-schema-parser/dist/src/directives.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var default_1 = require("./default");
exports.default = {
    // ** definition directives
    namespace: default_1.DefaultDirective,
    collection: default_1.DefaultDirective,
    version: default_1.DefaultDirective,
    extends: default_1.DefaultDirective,
    ignore: default_1.DefaultDirective,
    // ** control directives
    default: default_1.DefaultDirective,
    required: default_1.DefaultDirective,
    unique: default_1.DefaultDirective,
    setonce: default_1.DefaultDirective,
    uuid: default_1.DefaultDirective,
    optional: default_1.DefaultDirective,
    // ** math directives
    min: default_1.DefaultDirective,
    max: default_1.DefaultDirective,
    // ** string directives
    minlength: default_1.DefaultDirective,
    maxlength: default_1.DefaultDirective,
    trim: default_1.DefaultDirective,
    uppercase: default_1.DefaultDirective,
    lowercase: default_1.DefaultDirective,
    hash: default_1.DefaultDirective,
    mongo: default_1.DefaultDirective,
    fragment: default_1.DefaultDirective
};
