// ts-graphql-plugin/lib/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ErrorWithLocation extends Error {
    constructor(message, errorContent) {
        super(message);
        this.message = message;
        this.errorContent = errorContent;
        this.name = 'ErrorWithLocation';
        this.severity = 'Error';
        if (errorContent.severity) {
            this.severity = errorContent.severity;
        }
    }
}
exports.ErrorWithLocation = ErrorWithLocation;
class ErrorWithoutLocation extends Error {
    constructor(message, severity = 'Error') {
        super(message);
        this.message = message;
        this.severity = severity;
        this.name = 'ErrorWithoutLocation';
    }
}
exports.ErrorWithoutLocation = ErrorWithoutLocation;
exports.ERROR_CODES = {
    graphqlLangServiceError: {
        code: 51001,
    },
    templateIsTooComplex: {
        code: 51010,
        message: 'This operation or fragment has too complex interpolation to analize.',
    },
    errorInOtherInterpolation: {
        code: 51011,
        message: 'This expression has some GraphQL errors.',
    },
    schemaBuildError: {
        code: 51020,
    },
};
