// ts-graphql-plugin/lib/schema-manager/schema-manager-host.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
function createSchemaManagerHostFromPluginInfo(info) {
    return {
        getConfig() {
            return info.config;
        },
        fileExists(path) {
            return info.serverHost.fileExists(path);
        },
        readFile(path, encoding) {
            return info.serverHost.readFile(path, encoding);
        },
        watchFile(path, cb, interval) {
            return info.serverHost.watchFile(path, cb, interval);
        },
        getProjectRootPath() {
            return path_1.default.dirname(info.project.getProjectName());
        },
        log(msg) {
            info.project.projectService.logger.info(`[ts-graphql-plugin] ${msg}`);
        },
    };
}
exports.createSchemaManagerHostFromPluginInfo = createSchemaManagerHostFromPluginInfo;
